pkg/httperrors/httperrors.go

```diff
@@ -122,3 +123,22 @@ func HTTPError(c echo.Context, statusCode int, message string) (e error) {
 
 	return c.JSON(statusCode, resBody)
 }
+
+// GetHTTPResponseFromErr returns the mapped http error to the generic errors model.
+func GetHTTPResponseFromErr(c echo.Context, err error) (e error) {
+
+	switch err {
+	case models.ErrInternalServerError:
+		return InternalServerError(c, err.Error())
+	case models.ErrNotFound:
+		return NotFound(c, err.Error())
```

Comment:
Hi @davidffrench ,

Is the error "models.ErrNotFound" implemented to be returned when a **DATA** is NOT FOUND? If yes, the HTTP code 404 returned here according to the HTTP definitions should be used to indicates that the client (web browser) is receiving a message from the server (remote computer) that the specific resource (web page/URL) is unavailable. It is not a valid code when DATA in the database is not found/do not exist. 

Following the spec definition. 

"10.4.5 404 Not Found
The server has not found anything matching the Request-URI. No indication is given of whether the condition is temporary or permanent. The 410 (Gone) status code SHOULD be used if the server knows, through some internally configurable mechanism, that an old resource is permanently unavailable and has no forwarding address. This status code is commonly used when the server does not wish to reveal exactly why the request has been refused, or when no other response is applicable."

In this case, I understand that we could return 204. 

Reference: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html

Reply:
You have sparked an age-old debate which there never seems to be a definitive answer on. Here are some resources on the matter.

- https://stackoverflow.com/questions/11746894/what-is-the-proper-rest-response-code-for-a-valid-request-but-an-empty-data
- https://softwareengineering.stackexchange.com/questions/203492/when-to-use-http-status-code-404-in-an-api

Here is how I see it for our use case. We should probably be returning a 204 on /apps if there are no apps that are yet bound. We should be returning a 404 on /apps/{id} if that specific app does not exist.

Reply:
Changed the apps http handler to return a 204 now.

Reply:
Hi @davidffrench,

IHMO the nomenclature `NoContent` will be more appropriate too.

Reply:
Not for a generic error message in my opinion, NotFound is a much more widely known message. The purpose of the errors model is to have no relation to the http delivery layer. NotFound is more generic in my opinion.

Reply:
The point is that here will return 404 when a data does not exist in the database. For it returns the code 204 according to HTTP code definitions then is required to add [here](https://github.com/aerogear/mobile-security-service/blob/ecb03c5d1f8dfb515f844b85897ac04410a8cc69/pkg/httperrors/httperrors.go#L68) and the func called will be NoContent.

Reply:
@camilamacedo86 I completely agree that for the /apps endpoint a 204 is what makes the most sense to return. This is not the case for /apps/{id} which is a specific resource. As defined in the RFC for HTTP 1.1 (https://tools.ietf.org/html/rfc7231#page-59)

>   The 404 (Not Found) status code indicates that the origin server did
   not find a current representation for the target resource or is not
   willing to disclose that one exists.

A 204 (https://tools.ietf.org/html/rfc7231#section-6.3.5) could be interpreted to be used in this case but it should not be.

Reply:
HI @davidffrench,
 
Note that `current representation for the target resource and is not willing to disclose that one exists` is referent to `indicates that the origin server`. The target is the web server which means the endpoint. However, in the case of `/apps/{id}` it definitely could be applied since the ID is part of the endpoint then it also could mean that the endpoint does not exist. 

So, anyway, should not it here also have the 204 mapped since it is a generic http erros mapping?

Reply:
Thanks, Camila, I understand it could be interpreted as that. Our API will not be exposed for public use, we will not need to worry about this scenario.